elastic/kibana
# [Reporting/Cloud] Soft disable reporting functionality on 1GB Kibanas

## Summary

Initial work for soft-disabling rather than crashing Kibana when generating reports on Kibana instances with low available RAM. This work ensures that Kibana server responds with the appropriate messages via our reporting HTTP APIs when generating a report.

## How to test

1. Start Kibana & ES
2. Enable a trial license
3. Either test on Cloud deployment or, if testing locally, by applying this patch to simulate being on cloud on a 1GB instance
4. Generate a PDF or PNG report
5. See the error message in the UI
6. See the error message in the report info panel flyout that contains a link

## Screenshots

<img width="574" alt="Screenshot 2022-04-12 at 16 26 42" src="https://user-images.githubusercontent.com/8155004/162987395-3ff2328c-488e-462d-99ed-32b0a2480cc3.png">


## Implementation details and notes

* The `InsufficientMemoryAvailableOnCloudError` will be thrown on Cloud when we are running on a Kibana with <2GB RAM. Specifically, the error is thrown when we call `getScreenshots` exposed by the screenshotting plugin. In this way we block all screenshots from running.
* Reporting now knows about this error type and sets an appropriate message explaining what is wrong. Additionally, the UI shows a message with a link, guiding users to how the can get around the issue
* For `ReportingError` classes it is convenient to read `this.code` and `ClassName.code` I've opted to make `code` both an instance and static property name
* In collaboration with Cloud stakeholders, we have decided that reading `os.totalmem` is sufficient for informing the soft disable decision
* No new methods were added to screenhotting's API, although we may have to consider this if we want to pre-emptively disable the "generate report" button

## Related to

https://github.com/elastic/kibana/pull/129989


### Checklist

Delete any items that are not applicable to this PR.

- [ ] Test on cloud
